Add PGM-style allocation stack logging for bmalloc allocations

danlliu opened the pull request:
#### 771744e5b7906438f8f7864a59ffb6543f3aeaef
<pre>
Add PGM-style allocation stack logging for bmalloc allocations
<a href="https://bugs.webkit.org/show_bug.cgi?id=299621">https://bugs.webkit.org/show_bug.cgi?id=299621</a>
<a href="https://rdar.apple.com/153737681">rdar://153737681</a>

Reviewed by Yusuke Suzuki.

As another useful diagnostic tool, we want to add in allocation stack logging for
bmalloc&apos;s allocations. Memory Allocation Records (MAR) track metadata about where
allocations were made, enabling the analysis of these records if we crash. We aim
to utilize this to help debug bugs involving misuse of allocated memory, likely a
crash caught by features such as MIE. MAR ensures that the memory access patterns
are not disrupted, but we are still able to log some metadata to diagnose bugs at
crash time.

This patch was previously landed, but caused internal builds to fail. The updated
patch here fixes some module-related issues, which should fix these build issues.

Canonical link: <a href="https://commits.webkit.org/301840@main">https://commits.webkit.org/301840@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04cb16e4e9ae9c4c881f77c31a44e34f345168b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127222 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46870 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134286 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bc4695a3-4381-4275-ba30-fcbc7fcc7dd6) 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/47480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55391 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96833 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130170 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/47480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113944 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77332 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dca06ab0-bf1f-4439-b4aa-87fbb15bcbd9) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/47480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77666 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119259 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/47480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32438 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136769 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125685 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/53881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/41516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110295 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105037 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26772 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/50552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51444 "Built successfully") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/53818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59905 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158723 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39697 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/54811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/API/MARReportCrashPrivate.h:
```diff
@@ -0,0 +1,44 @@
+extern "C" {
+#endif
+
+JS_EXPORT kern_return_t MARReportCrashExtractResults(vm_address_t fault_address, mach_vm_address_t mar_global_registry, unsigned version, task_t, pas_mar_crash_report*, crash_reporter_memory_reader_t crm_reader);
```
Let's avoid exposing `struct pas_mar_crash_report` definition as it becomes ABI boundary. If we need to offer a way to access to the fields of this struct, let's define a getter / setter here.
And let's avoid including `pas_mar_report_crash.h` here.

Constellation commented on Source/JavaScriptCore/API/MARReportCrashPrivate.h:
```diff
@@ -0,0 +1,67 @@
+typedef struct OpaquePASMARCrashReport* PASMARCrashReportRef;
+
+JS_EXPORT kern_return_t MARReportCrashExtractResults(vm_address_t fault_address, mach_vm_address_t mar_global_registry, unsigned version, task_t, PASMARCrashReportRef, crash_reporter_memory_reader_t crm_reader);
+
```
Probably, `Create` and `Release` are necessary for `PASMARCrashReportRef`?

Constellation commented on Source/JavaScriptCore/API/MARReportCrashPrivate.cpp:
```diff
@@ -0,0 +1,121 @@
+
+PASMARCrashReportRef MARCrashReportCreate()
+{
+    return reinterpret_cast<PASMARCrashReportRef>(new pas_mar_crash_report);
```
Let's use `fastMalloc` explicitly as this `new` will allocate memory from the system malloc.
```
void* memory = fastMalloc(sizeof(pas_mar_crash_report));
return new (memory) pas_mar_crash_report;
```

Constellation replied:
Or, we can define a new struct which just wraps `pas_mar_crash_report`, and having a field for `pas_mar_crash_report`.

Constellation commented on Source/JavaScriptCore/API/MARReportCrashPrivate.cpp:
```diff
@@ -0,0 +1,121 @@
+
+void MARCrashReportRelease(PASMARCrashReportRef report)
+{
+    delete reinterpret_cast<pas_mar_crash_report*>(report);
```
And use `fastFree`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
```diff
@@ -79,6 +79,7 @@ struct pas_report_crash_pgm_report {
     pas_backtrace_metadata* dealloc_backtrace;
     bool pgm_has_been_used;
 };
+
```
Maybe, we can remove all changes for this header.

danlliu committed:
Add PGM-style allocation stack logging for bmalloc allocations
https://bugs.webkit.org/show_bug.cgi?id=299621
rdar://153737681

Reviewed by Yusuke Suzuki.

As another useful diagnostic tool, we want to add in allocation stack logging for
bmalloc's allocations. Memory Allocation Records (MAR) track metadata about where
allocations were made, enabling the analysis of these records if we crash. We aim
to utilize this to help debug bugs involving misuse of allocated memory, likely a
crash caught by features such as MIE. MAR ensures that the memory access patterns
are not disrupted, but we are still able to log some metadata to diagnose bugs at
crash time.

This patch was previously landed, but caused internal builds to fail. The updated
patch here fixes some module-related issues, which should fix these build issues.

Canonical link: https://commits.webkit.org/301840@main
